twitter_api/main.py

```diff
@@ -1,28 +1,217 @@
+    sql_command = 'SELECT * FROM tweet WHERE id = ?'
+    tweet_cursor = g.db.execute(sql_command, [tweet_id])
+    tweet = tweet_cursor.fetchone()
+    if not tweet:
+        abort(404)
+    return tweet
+    
+def _get_user_id_from_token(request):
+    # assumes access token has already been verified but can double check
+    if 'access_token' not in request.get_json():
+        abort(401)
+    access_token = request.get_json()['access_token']
+    sql_command = 'SELECT user_id FROM auth WHERE access_token = ?'
+    user_id_cursor = g.db.execute(sql_command, [access_token])
+    found_user = user_id_cursor.fetchone()
+    if not found_user:
+        abort(404)
+    return found_user[0]
+    
+def _convert_date_iso_8601(created):
+    dt_object = datetime.strptime(created, "%Y-%m-%d %H:%M:%S")
+    return dt_object.strftime("%Y-%m-%dT%H:%M:%S")
+    
+@app.route('/login', methods = ['POST'])
+@json_only
+def login():
+    data = request.get_json()
+    
+    if 'username' not in data:
+        abort(400) # Bad request, no username
+    elif 'password' not in data:
+        abort(400) # Bad request, no password
 
+    post_username = data['username']
+    post_password = data['password']
+    
+    sql_command = 'SELECT password, id FROM user WHERE username = ?'
+    
+    found_user_cursor = g.db.execute(sql_command, [post_username])
+    found_user = found_user_cursor.fetchone()
+    if not found_user:
+        abort(404) # Not found, username did find a match in db
+    password, id = found_user
+        
+    post_password = post_password.encode(request.charset) # note from Santiago
+    post_password = md5(post_password).hexdigest()
+    
+    if password != post_password:
+        abort(401) # Unauthorized
+        
+    # generate access token
+    char_list = string.ascii_lowercase + string.ascii_uppercase
+    access_token = ''.join(random.choice(char_list) for _ in range(10))
+    
+    sql_command = 'INSERT INTO auth (user_id, access_token) VALUES (?, ?)'
+    g.db.execute(sql_command, [id, access_token])
+    g.db.commit()
+    
+    return_dict = dict(access_token = access_token)
+    response = Response(
+        json.dumps(return_dict), status=201, content_type=JSON_MIME_TYPE)
+    return response
+    
+@app.route('/logout', methods = ['POST'])
+@json_only
+@auth_only
+def logout():
+    # Delete all entries for this user (they could have logged in multiple times)
+    # good suggestion by Santiago
+    user_id = _get_user_id_from_token(request)
+    sql_command = 'DELETE FROM auth WHERE user_id = ?'
+    g.db.execute(sql_command, [user_id])
+    g.db.commit()
+    return Response('', 204) # No Content
+    
+@app.route('/tweet/<tweet_id>')
+def get_tweet(tweet_id):
+    # get all the tweet info as well as username from user table
+    sql_command = '''SELECT t.*, u.username FROM tweet as t 
+                    INNER JOIN user as u ON t.user_id=u.id 
+                    WHERE t.id = ?'''
```

Comment:
Again a bit pedantic but be careful with your SQL here. As you are using an ```inner``` join if the tweet exists but the user has since been deleted then your SQL query will return nothing. Better to use a ```left``` join:
```SQL
SELECT t.*, u.username FROM tweet t LEFT JOIN user u ON t.user_id = u.id WHERE t.id = ?
```